docs/streams/developer-guide.html

```diff
@@ -267,21 +295,532 @@
+               stream.foreach((key, value) -> System.out.println(key + " => " + value));
+
+               // Java 7 example
+               stream.foreach(
+                   new ForeachAction&lt;String, Long&gt;() {
+                       @Override
+                       public void apply(String key, Long value) {
+                         System.out.println(key + " => " + value);
+                       }
+               });
+               </pre>
+           </td>
+       </tr>
+       <tr>
+           <td><b>GroupByKey</b>: <code>KStream->KGroupedStream </code></td>
+           <td>
+               Groups the records by the existing key.<p>
+
+               Grouping is a prerequisite for aggregating a stream or a table and ensures that data is properly
+               partitioned ("keyed") for subsequent operations.<p>
+
+               <b>When to set explicit serdes</b>: Variants of <code>groupByKey</code> exist to override the configured default serdes of
+               your application, which you must do if the key and/or value types of the resulting <code>KGroupedStream</code> do
+               not match the configured default serdes.<p>
+
+               <b>Note:</b>
+               Grouping vs. Windowing: A related operation is windowing, which lets you control how to "sub-group" the
+               grouped records of the same key into so-called windows for stateful operations such as windowed aggregations
+               or windowed joins.<p>
+
+               Causes data re-partitioning if and only if the stream was marked for re-partitioning. <code>groupByKey</code> is
+               preferable to <code>groupBy</code> because it re-partitions data only if the stream was already marked for re-partitioning.
+	       However, <code>groupByKey</code> does not allow you to modify the key or key type like <code>groupBy</code> does.
+
+               <pre class="brush: java;">
+               KStream&lt;byte[], String&gt; stream = ...;
+
+               // Group by the existing key, using the application's configured
+               // default serdes for keys and values.
+               KGroupedStream&lt;byte[], String&gt; groupedStream = stream.groupByKey();
+
+               // When the key and/or value types do not match the configured
+               // default serdes, we must explicitly specify serdes.
+               KGroupedStream&lt;byte[], String&gt; groupedStream = stream.groupByKey(
+                   Serdes.ByteArray(), /* key */
+                   Serdes.String()     /* value */
+               );
+               </pre>
+           </td>
+       </tr>
+       <tr>
+           <td><b>GroupBy</b>: <code>KStream->KGroupedStream or KTable->KGroupedTable</code></td>
+           <td>
+               Groups the records by a new key, which may be of a different key type. When grouping a table,
+               you may also specify a new value and value type. groupBy is a shorthand for selectKey(...).groupByKey().<p>
```

Comment:
"groupBy is a shorthand for selectKey(...).groupByKey()." should be properly code-formatted:

> `groupBy` is a shorthand for `selectKey(...).groupByKey()`.